src/Confluent.Kafka/Impl/SafeKafkaHandle.cs

```diff
@@ -2460,12 +2472,61 @@ internal void AlterUserScramCredentials(IEnumerable<UserScramCredentialAlteratio
                 {
                     Librdkafka.UserScramCredentialAlteration_destroy(c_alterationsPtr[i]);
                 }
+            }
+        }
+
+        internal void ListOffsets(IEnumerable<TopicPartitionOffsetSpec> requests, ListOffsetsOptions options, IntPtr resultQueuePtr, IntPtr completionSourcePtr)
+        {
+            ThrowIfHandleClosed();
+            var optionsPtr = IntPtr.Zero;
+            var topic_partition_list = IntPtr.Zero;
+            try
+            {
+                // set Admin Options if any
+                options = options ?? new ListOffsetsOptions();
+                optionsPtr = Librdkafka.AdminOptions_new(handle, Librdkafka.AdminOp.ListOffsets);
+                setOption_RequestTimeout(optionsPtr, options.RequestTimeout);
+                setOption_IsolationLevel(optionsPtr, options.IsolationLevel);
+                setOption_completionSource(optionsPtr, completionSourcePtr);
+
+                topic_partition_list = Librdkafka.topic_partition_list_new((IntPtr)requests.Count());
+                foreach(var request in requests){
+                    string Topic = request.TopicPartition.Topic;
+                    int Partition = request.TopicPartition.Partition;
+                    IntPtr topic_partition = Librdkafka.topic_partition_list_add(topic_partition_list,Topic,Partition);
+                    var tp = Util.Marshal.PtrToStructure<rd_kafka_topic_partition>(topic_partition);
+                    if (request.OffsetSpec is EarliestOffsetSpec)
```

Comment:
With nested class, this can be replaced with `OffsetSpec.EarliestSpec` and similarly below.